src/os/bluestore/BlueStore.cc

```diff
@@ -5363,706 +5225,466 @@ void BlueStore::_pad_zeros_tail(
-	assert(cow_length == 0);  // fixme ... until we fix this loop
+      if (tail_read) {
+	bufferlist tail_bl;
+	int r = _do_read(c.get(), o, offset + length + tail_pad, tail_read,
+			 tail_bl, 0);
+	assert(r >= 0 && r <= (int)tail_read);
+	b_len += r;
+	padded.claim_append(tail_bl);
+	size_t zlen = tail_read - r;
+	if (zlen) {
+	  padded.append_zero(zlen);
+	}
       }
-    } else {
-      dout(20) << "  tail shared, but no COW needed" << dendl;
     }
-    if (cow_offset_raw < o->onode.size &&
-	(cow_offset_raw & ~block_mask)) {
-      *cow_rmw_tail = b->calc_offset(cow_offset_raw - bp->first, nullptr);
-      dout(20) << "  cow_rmw_tail 0x" << std::hex << *cow_rmw_tail << std::dec
-	       << " from " << bp->second
-	       << " blob " << *b << dendl;
+
+    // chunk-aligned wal overwrite?
+    if (b->get_ondisk_length() >= b_off + b_len &&
+	b_off % chunk_size == 0 &&
+	b_len % chunk_size == 0) {
+      bluestore_wal_op_t *op = _get_wal_op(txc, o);
+      op->op = bluestore_wal_op_t::OP_WRITE;
+      b->map(b_off, b_len,
+	     [&](uint64_t offset, uint64_t length) {
+	       op->extents.emplace_back(bluestore_pextent_t(offset, length));
+	     });
+      if (b->csum_type) {
+	checksummer->calculate(b->csum_type, b->get_csum_block_size(),
+			       b_off, padded.length(), padded, &b->csum_data);
+      }
+      op->data.claim(padded);
+      dout(20) << __func__ << "  wal write 0x" << std::hex << b_off << "~0x"
+	       << b_len << std::dec << " of mutable " << blob << ": " << *b
+	       << " at " << op->extents << dendl;
+      o->onode.punch_hole(offset, length, &wctx->lex_old);
+      bluestore_lextent_t& lex = o->onode.extent_map[offset] =
+	bluestore_lextent_t(blob, offset - bstart, length, 0);
+      b->ref_map.get(lex.offset, lex.length);
+      dout(20) << __func__ << "  lex 0x" << std::hex << offset
+	       << std::dec << ": " << lex << dendl;
+      dout(20) << __func__ << "  old " << blob << ": " << *b << dendl;
+      return;
     }
+
+    ++ep;
   }
 
-  if (length) {
-    dout(20) << "  must alloc 0x" << std::hex << offset << "~0x" << length
-	     << std::dec << dendl;
+  // new blob.
+  int64_t blob;
+  b = o->onode.add_blob(&blob);
+  b->length = min_alloc_size;
+  uint64_t b_off = offset % min_alloc_size;
+  uint64_t b_len = length;
+  _pad_zeros(txc, o, &bl, &b_off, &b_len, block_size);
+  o->onode.punch_hole(offset, length, &wctx->lex_old);
+  bluestore_lextent_t& lex = o->onode.extent_map[offset] =
+    bluestore_lextent_t(blob, offset % min_alloc_size, length);
+  b->ref_map.get(lex.offset, lex.length);
+  // it's little; don't bother compressing
+  b->set_flag(bluestore_blob_t::FLAG_MUTABLE);
+  if (csum_type) {
+    // it's little; csum at block granularity.
+    b->init_csum(csum_type, block_size_order, min_alloc_size);
+    checksummer->calculate(b->csum_type, b->get_csum_block_size(),
+			   b_off, b_len, bl, &b->csum_data);
+  }
 
-    // positional hint
-    uint64_t hint = 0;
+  // allocate and write
```

Comment:
IMHO alloc/write code below duplicates one from do_alloc_write - why not reuse it here?